GA4GH data source

danvk opened the pull request:
As demoed last Friday.

Highlights:
- I factored out a common interface for alignments, shared between BAM & GA4GH sources.
- GA4GHDataSource is currently untested. I'm thinking about how to best remedy that.
- The BAM↔GA4GH equivalency test surfaced many issues with the GA4GH server.
- This isn't surfaced via the playground example. I'm not sure how it could be (you'd have to run a GA4GH server). Maybe there's a public one running that we could point to.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/hammerlab/pileup.js/230)
<!-- Reviewable:end -->

armish commented:
looks great, Dan! Looking forward to seeing this in action.

I just added a few minor comments to point out pieces that confused me, but feel free to ignore them if they don't make sense.

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->

armish commented:
Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[src/main/Alignment.js, line 7 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7Sa3nSguwvqpJEWPo)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/Alignment.js#L7)):</sup>
do you think it makes sense (for people like me who don't know anything about this format) to add a comment/pointer here that provides more context for the CIGAR terminology? 

---

<sup>**[src/main/Alignment.js, line 9 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7SCWoddhwLGIRRrLQ)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/Alignment.js#L9)):</sup>
is there a way/test to make sure that every Alignment implementation covers these options? I can easily imagine myself missing one of those options in the future and debugging to see what went wrong :)

---

<sup>**[src/main/SamRead.js, line 113 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7Tw7WoWmurr2vR6xo)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/SamRead.js#L113)):</sup>
sorry if I am missing an apparent point here, but why do we need `_interval` again? Does this act like a state variable for caching purposes?

---

<sup>**[src/main/SamRead.js, line 117 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7QVylmB3xzOF-GcHf)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/SamRead.js#L117)):</sup>
is this a left-over line?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->

danvk commented:
Addressed all comments, let me know if it's good to merge.

---

Review status: 8 of 10 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

<sup>**[src/main/Alignment.js, line 7 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7Sa3nSguwvqpJEWPo)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/Alignment.js#L7)):</sup>
If this was confusing to you then it will be confusing to others, too. I added a note.

---

<sup>**[src/main/Alignment.js, line 9 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7SCWoddhwLGIRRrLQ)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/Alignment.js#L9)):</sup>
Flow gains support for disjoint unions in v0.13.1, which can be used to [model an enum type][1]. We could give them a try here.

[1]: https://github.com/facebook/flow/issues/627

---

<sup>**[src/main/SamRead.js, line 113 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7Tw7WoWmurr2vR6xo)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/SamRead.js#L113)):</sup>
Yep. I added a comment. It would be a good idea to evaluate whether this memory/speed tradeoff actually makes sense.

---

<sup>**[src/main/SamRead.js, line 117 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7QVylmB3xzOF-GcHf)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/SamRead.js#L117)):</sup>
Yes, thanks for the catch!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->

armish commented:
yeah, all looks good -- thanks for those clarifications, Dan!

---







Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->

armish commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

<sup>**[src/main/Alignment.js, line 9 \[r1\]](https://reviewable.io:443/reviews/hammerlab/pileup.js/230#-Ju7SCWoddhwLGIRRrLQ)** ([raw file](https://github.com/hammerlab/pileup.js/blob/fb8b765ecf5a0a5e27ccd388e6a20fa465618d4d/src/main/Alignment.js#L9)):</sup>
that would be great to have -- I will keep an eye on this.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->

danvk commented:
Hmm, "Coverage decreased (-2.78%) to 92.34%" → do not like. Lemme see what I can come up with to improve this.

armish commented:
LTGM.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/hammerlab/pileup.js/230)
<!-- Sent from Reviewable.io -->
